apollo-federation/src/sources/connect/json_selection/README.md

```diff
@@ -55,7 +55,23 @@ improvements, we should adhere to the following principles:
    avoided because it limits the developer's ability to subselect fields of the
    opaque `JSON` value in GraphQL operations.
 
-3. Backwards compatibility should be maintained as we release new versions of
+3. `JSONSelection` syntax may be subsetted arbitrarily, either by generating a
+   reduced `JSONSelection` that serves the needs of a particular GraphQL
+   operation, or by skipping unneeded selections during `ApplyTo` execution.
+   When this subsetting happens, it would be highly undesirable for the behavior
+   of the remaining selections to change unexpectedly.
+
+   For example, it is simpler to insist that `{...}` selections are always
+   mapped over input arrays, rather than sometimes mapping (when named fields
+   are present, say) and sometimes applying directly to the array itself (when
+   only array-safe `$->method` selections are present, hypothetically). Even if
+   array safety was easy to decide, it would be a mistake to base the mapping
+   decision on the presence or absence of certain kinds of selections, because
+   then the mapping behavior could change unexpectedly whenever certain kinds of
+   selections happen to be omitted, which is an ever-present possibility thanks
+   to subsetting.
```

Comment:
What about striking this whole paragraph? Hopefully the previous paragraph about subsetting consistency says enough already?

3. `JSONSelection` syntax may be subsetted arbitrarily, either by generating a
   reduced `JSONSelection` that serves the needs of a particular GraphQL
   operation, or by skipping unneeded selections during `ApplyTo` execution.
   When this subsetting happens, it would be highly undesirable for the behavior
   of the remaining selections to change unexpectedly.

We're still pretty early in the document, so brevity seems worthwhile. We don't need to justify the principle here.

Reply:
Works for me! We could also end up talking about function purity / side effects as a more general rule to prevent subsetting mischief

Reply:
https://github.com/apollographql/router/pull/5762/commits/a0c5d92bddf9ce414cc263c62eb730f74b0659e2